gitlab/resource_gitlab_project_protected_environment.go

```diff
@@ -0,0 +1,215 @@
+		acceptedAccessLevels = append(acceptedAccessLevels, k)
+	}
+	return &schema.Resource{
+		Create: resourceGitlabProjectProtectedEnvironmentCreate,
+		Read:   resourceGitlabProjectProtectedEnvironmentRead,
+		Delete: resourceGitlabProjectProtectedEnvironmentDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"environment": {
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"deploy_access_levels": {
+				Type:     schema.TypeList,
+				ForceNew: true,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"access_level": {
+							Type:         schema.TypeString,
+							ForceNew:     true,
+							Optional:     true,
+							ValidateFunc: validateValueFunc(acceptedAccessLevels),
+						},
+						"access_level_description": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"user_id": {
+							Type:         schema.TypeInt,
+							ForceNew:     true,
+							Optional:     true,
+							ValidateFunc: validation.IntAtLeast(1),
+						},
+						"group_id": {
+							Type:         schema.TypeInt,
+							ForceNew:     true,
+							Optional:     true,
+							ValidateFunc: validation.IntAtLeast(1),
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceGitlabProjectProtectedEnvironmentCreate(d *schema.ResourceData, meta interface{}) error {
+	deployAccessLevels, err := expandDeployAccessLevels(d.Get("deploy_access_levels").([]interface{}))
+	if err != nil {
+		return fmt.Errorf("error expanding deploy_access_levels: %v", err)
+	}
+	environment := d.Get("environment").(string)
+	options := gitlab.ProtectRepositoryEnvironmentsOptions{
+		Name:               &environment,
+		DeployAccessLevels: deployAccessLevels,
+	}
+
+	project := d.Get("project").(string)
+
+	log.Printf("[DEBUG] Project %s create gitlab protected environment %q", project, *options.Name)
+
+	client := meta.(*gitlab.Client)
+
+	protectedEnvironment, resp, err := client.ProtectedEnvironments.ProtectRepositoryEnvironments(project, &options)
+	if err != nil {
+		if resp.StatusCode == http.StatusNotFound {
+			return fmt.Errorf("feature Protected Environments is not available")
+		}
+		return err
+	}
+
+	d.SetId(buildTwoPartID(&project, &protectedEnvironment.Name))
+
+	return resourceGitlabProjectProtectedEnvironmentRead(d, meta)
+}
+
+func resourceGitlabProjectProtectedEnvironmentRead(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] read gitlab protected environment %s", d.Id())
+
+	project, environment, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return err
+	}
+	d.Set("project", project)
+	d.Set("environment", environment)
```

Comment:
Would be good to handle `error` for `Set`:

```suggestion
	err = d.Set("project", project)
	if err != nil {
		return err
	}

	err = d.Set("environment", environment)
	if err != nil {
		return err
	}
```